docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
 querying. However, this ADR solely focuses on transactions. Querying should be
 addressed in a future ADR, but it should build off of these proposals.
 
+Based on detailed discussions ([\#6030](https://github.com/cosmos/cosmos-sdk/issues/6030)
+and [\#6078](https://github.com/cosmos/cosmos-sdk/issues/6078)), the original 
+design for transactions was changed substantially from an `oneof` /JSON-signing
+approach to the approach described below.
+
 ## Decision
 
 ### Transactions
 
-Since the messages that an application is known and allowed to handle are specific
-to the application itself, so must the transactions be specific to the application
-itself. Similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
-the concrete types will be defined at the application level via Protobuf `oneof`.
+Since interface values are encoded with `google.protobuf.Any` in state (see [ADR 019](adr-019-protobuf-state-encoding.md)),
+ `sdk.Msg`s are encoding with `Any` in transactions.
+ 
+ One of the main goals of using `Any` to encode interface values is to have a
+ core set of types which is reused by apps so that
+ clients can safely be compatible with as many chains as possible.
+ 
+ It is one of the goals of this specification to provide a flexible cross-chain transaction
+ format that can serve a wide variety of use cases without breaking client 
+ compatibility.
+
+In order to facilitate signing, transactions are separated into `TxBody`,
+which will be re-used by `SignDoc` below, and `signatures`:
```

Comment:
The motivation is 1) it makes it easy to do signatures with raw bytes as described later, 2) just reusing the same struct on `SignDoc` rather than redefining the fields but possibly with different field numbers just seems more elegant

Reply:
@alexanderbez wonder what your thoughts are about sharing `TxBody` between `Tx` and `SignDoc`?

Reply:
I don't have a strong preference here :) embedding saves us the potential for missing fields by accident.

Reply:
The fields in here are not too many and are not expected to change often (in contract to the internals of the messages). If possible, I'd keep this flat for simplicity.

Reply:
As I've articulated elsewhere not using `TxBody` and instead going with a flat structure forces redefining the signing mechanism as being based on canonicalization rather than byte equality. That is undesirable if we want to introduce an extensibility mechanism that allows newer clients to interact with older servers (see https://github.com/cosmos/cosmos-sdk/pull/6111#discussion_r421025057)